scripts/run-alphanet-relay.sh

```diff
@@ -24,16 +29,40 @@ do
     RELAY_INDEX=$((RELAY_INDEX + 1))
     RELAY_PORT=$((RELAY_PORT + 100))
     
-    if [ $RELAY_PORT -ge $((USER_PORT + 1000)) ]
+    if [ $RELAY_PORT -ge $((USER_PORT + 300)) ]
     then
-        echo "No more relay port available! (limited to 9 relays)"
+        echo "No more relay port available! (limited to 3 relays)"
         exit 1
     fi
 done
 
 
+
 echo "relay ${RELAY_INDEX} - p2p-port: $((RELAY_PORT)), http-port: $((RELAY_PORT + 1)) , ws-port: $((RELAY_PORT + 2))"
 
+# This part will insert the keys in the node
+bash -c "sleep 5; \
+insertKey() { \
+    polkadot-js-api \
+        --ws 'ws://localhost:$((RELAY_PORT + 2))' \
+        --sudo \
+        --seed "$SUDO_SEED" \
+        rpc.author.insertKey \
+            \"\$1\" \
+            \"\$2\" \
+            \"\$3\" > /dev/null; \
+}; \
+\
+insertKey acco '${RELAY_SEEDS[$RELAY_INDEX]}' '${RELAY_SR25519_PUB[$RELAY_INDEX]}'; \
+insertKey stak '${RELAY_SEEDS[$RELAY_INDEX]}' '${RELAY_SR25519_PUB[$RELAY_INDEX]}'; \
+insertKey babe '${RELAY_SEEDS[$RELAY_INDEX]}' '${RELAY_SR25519_PUB[$RELAY_INDEX]}'; \
+insertKey gran '${RELAY_SEEDS[$RELAY_INDEX]}' '${RELAY_ED25519_PUB[$RELAY_INDEX]}'; \
```

Comment:
The reason finalization isn't working is that after you insert the grandpa key, you need to restart the node for it to take affect. This is a known issue https://github.com/paritytech/substrate/issues/7268

Unfortunately it probably complicates this script.

Reply:
Thank you for the explanation.